File: MtxVec.Delphi/DefaultArray.pas
unit DefaultArray;

{$I BdsppDefs.inc}
interface

uses
  Windows, Messages, SysUtils, Classes, Graphics, Controls, Forms, Dialogs,
  Basic2, StdCtrls, ExtCtrls, MtxVec, Math387, ComCtrls, TeEngine, Series,
  TeeProcs, Chart, Buttons, MtxExpr;

type
  TDefArray = class(TBasicForm2)
    TrackBar1: TTrackBar;                                 
    Label1: TLabel;
    Label2: TLabel;
    RadioGroup1: TRadioGroup;
    Button1: TBitBtn;
    Chart1: TChart;                                              
    Series1: TLineSeries;
    Series2: TLineSeries;
    Series3: TLineSeries;
    procedure FormCreate(Sender: TObject);
    procedure TrackBar1Change(Sender: TObject);
    procedure Button1Click(Sender: TObject);
    procedure RadioGroup1Click(Sender: TObject);
  private
    TestVec : Vector;
    TestMtx : Matrix;
    Len     : Integer;
    function CumulativeSum(A: TVec): TSample; overload;
    function CumulativeSum(A: TMtx): TSample; overload;
    function CumulativeSumFast1(A: TVec): TSample; overload;    
    function CumulativeSumFast(A: TVec): TSample;overload;
    procedure CumulativeSumFast(A: TMtx);overload;
    procedure CumulativeSumFast1(A: TMtx); overload;
    { Private declarations }
  public
    { Public declarations }
  end;

var
  DefArray: TDefArray;

implementation

uses AbstractMtxVec, Types, MtxVecBase;

{$IFDEF CLR}
{$R *.NFM}
{$ELSE}
{$R *.DFM}
{$ENDIF}

function TDefArray.CumulativeSumFast(A: TVec): TSample;
var i : Integer;
begin
     {$IFDEF TTDOUBLE}
     for i := 0 to A.Length - 1 do
        A.Values[i] := A.Values[i] + A.Values[i];
     {$ELSE}
     for i := 0 to A.Length - 1 do
        A.SValues[i] := A.SValues[i] + A.SValues[i];
     {$ENDIF}
     Result := A[0];
end;

function TDefArray.CumulativeSumFast1(A: TVec): TSample;
var i : Integer;
    a1 : TSampleArray;
begin
     Enlist(A,a1);
     for i := 0 to A.Length - 1  do
     begin
          a1[i] := a1[i] + a1[i];
     end;
     Result := a1[0];

     DismissIt(a,a1);
end;                             

function TDefArray.CumulativeSum(A: TVec): TSample;
var i : Integer;
begin
     for i := 0 to A.Length - 1 do
     begin
          A[i] := A[i] + A[i];
     end;
     Result := A[0];
end;

function TDefArray.CumulativeSum(A: TMtx): TSample;
var i,j : Integer;
begin
     for i := 0 to A.Rows-1 do
        for j := 0 to A.Cols-1 do
        begin
            A[i,j] := A[i,j] + A[i,j];
        end;
     Result := A[0,0];
end;

procedure TDefArray.CumulativeSumFast1(A: TMtx);
var i,j : Integer;
    a1  : T2DDoubleArray;
    sa1  : T2DSingleArray;
begin
     if a.IsDouble then
     begin
         Enlist(A,a1);

         for i := 0 to A.Rows-1 do
            for j := 0 to A.Cols-1 do
                a1[i,j] := a1[i,j] + a1[i,j];

         DismissIt(A,a1);
     end else
     begin
         Enlist(A, sa1);

         for i := 0 to A.Rows-1 do
            for j := 0 to A.Cols-1 do
                sa1[i,j] := sa1[i,j] + sa1[i,j];

         DismissIt(A, sa1);
     end;
end;

procedure TDefArray.CumulativeSumFast(A: TMtx);
var i,j  : Integer;
    k, Cols  : Integer;
begin
     Cols := A.Cols;
     {$IFDEF TTDOUBLE}
     for i := 0 to A.Rows-1 do
        for j := 0 to A.Cols-1 do
        begin
            k := i*Cols+j;
            A.Values1D[k] := A.Values1D[k] + A.Values1D[k];
        end;
     {$ELSE}
     for i := 0 to A.Rows-1 do
        for j := 0 to A.Cols-1 do
        begin
            k := i*Cols+j;
            A.SValues1D[k] := A.SValues1D[k] + A.SValues1D[k];
        end;
     {$ENDIF}
end;


procedure TDefArray.FormCreate(Sender: TObject);
begin
  inherited;
  With RichEdit1 do
  begin
    Lines.Clear;
    Lines.Add('There are three different ways on how to'
                      + ' access the elements of an array. '
                      + ' (vector or matrix). All access methods '+
                      '  work on the same block of memory.');

    Lines.Add('');
    Lines.Add('(*) Default arrays give you clear syntax, '
                      + 'but are not as fast:');
    SelAttributes.Name := 'Courier New';
    Lines.Add('a[i,j] := 1; {where "a" is TMtx object}');
    Lines.Add('');
    Lines.Add('(*) Dynamic array pointer is a bit faster than '
            + ' default array.');
    SelAttributes.Name := 'Courier New';
    Lines.Add('a.Values[i,j] := 3.5; {where "a" is TMtx object}');
    Lines.Add('');
    Lines.Add('(*) One dimensional array (TMtx object only) allows '
            + 'you to work with the matrix as you would with one '
            + 'dimensional array:');
    SelAttributes.Name := 'Courier New';
    Lines.Add('a.Values1D[i*Cols+j]:= 2.3; {where "a" is TMtx object}');
  end;
  RadioGroup1Click(RadioGroup1);
  TrackBar1Change(TrackBar1);
end;

procedure TDefArray.TrackBar1Change(Sender: TObject);
begin
  Len := TTrackBar(Sender).Position;
  Label1.Caption := IntToStr(Len);
end;

procedure TDefArray.Button1Click(Sender: TObject);
var i : Integer;
    j : Integer;
    MaxIter : Integer;
begin
  { setup TestVec }
  Series1.Clear;
  Series2.Clear;
  Series3.Clear;
  Screen.Cursor := crHourGlass;
  Series1.Title := 'Values property';
  Series2.Title := 'Values array field';
  Series3.Title := 'array of double';    
  if RadioGroup1.ItemIndex = 0 then { test vector }
  begin
    j := 1;
    MaxIter := 10;
    if Len < 5000 then MaxIter := 500;
    if Len < 2000 then MaxIter := 1000;
    if Len < 500 then MaxIter := 6000;
    if Len < 100 then MaxIter := 30000;
    while j < Len do
    begin
      TestVec.Size(j);

      { default arrays }
      TestVec.SetZero;
      StartTimer;
      for i := 1 to MaxIter do CumulativeSum(TestVec);
      Series1.AddXY(j,StopTimer*1000,'',clTeeColor);


      { TSampleArray }
      TestVec.SetZero;
      StartTimer;
      for i := 1 to MaxIter do CumulativeSumFast(TestVec);
      Series2.AddXY(j,StopTimer*1000,'',clTeeColor);

     { TSampleArray Enlist}
      TestVec.SetZero;
      StartTimer;
      for i := 1 to MaxIter do CumulativeSumFast1(TestVec);
      Series3.AddXY(j,StopTimer*1000,'',clTeeColor);


      if j < 10 then j := j +2
      else if J < 100 then j := j + 10
      else if j < 500 then j := j + 50
      else if j < 1000 then j := j + 100
      else if j < 3000 then j := j + 250
      else j := j + 1000;
    end
  end else
  begin
    j := 5;
    MaxIter := 1;
    if Len < 70 then MaxIter := 150;
    if Len < 30 then MaxIter := 550;

    while j < Len do
    begin
      TestMtx.Size(j,j);

      { default arrays }
      TestMtx.SetZero;
      StartTimer;
      for i := 1 to MaxIter do CumulativeSum(TestMtx);
      Series1.AddXY(j,StopTimer*1000,'',clTeeColor);

      { TSampleArray fast}
      StartTimer;
      for i := 1 to MaxIter do CumulativeSumFast(TestMtx);
      TimeElapsed := StopTimer;
      Series2.AddXY(j,StopTimer*1000,'',clTeeColor);

      { TSampleArray ultra fast}
      StartTimer;
      for i := 1 to MaxIter do CumulativeSumFast1(TestMtx);
      TimeElapsed := StopTimer;
      Series3.AddXY(j,StopTimer*1000,'',clTeeColor);

      if j < 10 then j := j +2
      else if j < 50 then j := j + 5
      else if j < 100 then j := j + 10
      else if j < 200 then j := j + 15
      else j := j + 20;
    end;
  end;
  Chart1.LeftAxis.Title.Caption := 'time [ms] ('+IntToStr(MaxIter)+' iterations)';
  Screen.Cursor := crDefault;
end;

procedure TDefArray.RadioGroup1Click(Sender: TObject);
begin
  Case TRadioGroup(Sender).ItemIndex of
  0  : begin
            Label2.Caption := 'Vector length';
            TrackBar1.Max := 10000;
            TrackBar1.Frequency := 500;
       end;
  1  : begin
            Label2.Caption := 'Sqared matrix rows';
            TrackBar1.Max := 300;
            TrackBar1.Frequency := 10;
            TrackBar1.Position := 150;
       end;
  end;
  TrackBar1Change(TrackBar1);
end;

initialization
   RegisterClass(TDefArray);

end.

File: MtxVec.CPP/DefaultArray.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "DefaultArray.h"
#include "MtxExpr.hpp"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "Basic2"
#pragma resource "*.dfm"
TDefArray *DefArray;
//---------------------------------------------------------------------------
__fastcall TDefArray::TDefArray(TComponent* Owner)
  : TBasicForm2(Owner)
{
    AnsiString str;

    RichEdit1->Lines->Clear();

	str = "There are three different ways on how to";
	str+= " access the elements of an array. ";
	str+= " (vector or matrix). All access methods ";
	str+= "  work on the same block of memory. Include MtxExpr.hpp to gain access to sVector, cVector, sMatrix and cMatrix types.";

	RichEdit1->Lines->Add("");
	str = "(*) Default arrays with ( , ) operators give you a clear syntax:";
	RichEdit1->Lines->Add(str);
	RichEdit1->SelAttributes->Name = "Courier New";
	RichEdit1->Lines->Add("a.Values(i , j) = 1; {where ""a"" is sMatrix object}");
	RichEdit1->Lines->Add("a.CValues(i , j) = Cplx(1,2)");
	RichEdit1->Lines->Add("");
	RichEdit1->Lines->Add("(*) Indexed pointers are fast as always:");
	RichEdit1->SelAttributes->Name = "Courier New";
	RichEdit1->Lines->Add("double *ap; ");
	RichEdit1->Lines->Add("ap = a.PValues1D(0)"); //obtain a pointer to 0th element
	RichEdit1->Lines->Add("ap = a.PValues(Row,Col)"); //obtain a pointer to (Row*a->Cols+Col) th element
	RichEdit1->Lines->Add("ap[i] = 3.5; {where ""a"" is TMtx object}");
	RichEdit1->Lines->Add("");
	str = "(*) One dimensional array (TMtx object only) allows ";
	str+= "you to work with the matrix as you would with one ";
	str+= "dimensional array:";
	RichEdit1->Lines->Add(str);
	RichEdit1->SelAttributes->Name = "Courier New";
	RichEdit1->Lines->Add("am[i*Cols+j] = 2.3; ");
	RichEdit1->Lines->Add("");
	RichEdit1->Lines->Add("For vectors");
	RichEdit1->Lines->Add("av[i] = 2.3; ");

	RadioGroup1Click(RadioGroup1);
	TrackBar1Change(TrackBar1);
}

//---------------------------------------------------------------------------
void __fastcall TDefArray::CumulativeSum(sVector &A)
{
  if (A.IsDouble) {

	  for (int i=0; i < A.Length ;i++)
		   A[i] = A[i] + A[i];
  }
  else
  {
	  for (int i=0; i < A.Length ;i++)
		   A.SValues(i) = A.SValues(i) + A.SValues(i);
  }

}

//---------------------------------------------------------------------------
void __fastcall TDefArray::CumulativeSumFast(sVector& A)
{
  if (A.IsDouble) {

	  double *ap;
	  ap = A.PValues1D(0);

	  for (int i=0; i < A.Length ;i++) {
			ap[i] = ap[i] + ap[i];
	  }
  }
  else
  {
	  float *ap;
	  ap = A.PSValues1D(0);

	  for (int i=0; i < A.Length ;i++) {
			ap[i] = ap[i] + ap[i];
	  }

  }
}
//---------------------------------------------------------------------------
void __fastcall TDefArray::CumulativeSum(sMatrix& A)
{                                    //Values 2D property
  if (A.IsDouble) {

	  for (int i=0; i < A.Rows ;i++)
		for (int j=0; j < A.Cols ;j++)
		  A.Values(i,j) = A.Values(i,j) + A.Values(i,j);
  }
  else
  {
	  for (int i=0; i < A.Rows ;i++)
		for (int j=0; j < A.Cols ;j++)
		  A.SValues(i,j) = A.SValues(i,j) + A.SValues(i,j);
  }
}
//---------------------------------------------------------------------------
void __fastcall TDefArray::CumulativeSumFast(sMatrix& A)
{                             // Indexed pointer
  int Cols = A.Cols;
  int k,i,j;

  if (A.IsDouble) {

	  double *ap;
	  ap = A.PValues1D(0);

	  for (i=0; i < A.Rows ;i++)
		for (j=0; j < A.Cols ;j++) {
		  k = i*Cols+j;
		  ap[k]= ap[k] + ap[k];
	  }

  } else {

	  float *ap;
	  ap = A.PSValues1D(0);

	  for (i=0; i < A.Rows ;i++)
		for (j=0; j < A.Cols ;j++) {
		  k = i*Cols+j;
		  ap[k]= ap[k] + ap[k];
	  }

  }

}
//---------------------------------------------------------------------------
void __fastcall TDefArray::CumulativeSumFast1(sMatrix& A)
{                                  // Indexed operator
  int Cols = A.Cols;
  int k,i,j;

  if (A.IsDouble) {

	  for (i=0; i < A.Rows ;i++)
		for (j=0; j < A.Cols ;j++) {

		  k = i*Cols+j;
		  A[k] = A[k] + A[k];
		}

  } else {

	  for (i=0; i < A.Rows ;i++)
		for (j=0; j < A.Cols ;j++) {

		  k = i*Cols+j;
		  A->SValues1D[k] = A->SValues1D[k] + A->SValues1D[k];
		}

  }

}
//---------------------------------------------------------------------------
void __fastcall TDefArray::TrackBar1Change(TObject *Sender)
{
  Len = TrackBar1->Position;
  Label1->Caption = IntToStr(Len);
}
//---------------------------------------------------------------------------

void __fastcall TDefArray::Button1Click(TObject *Sender)
{
  int i, j, MaxIter;

//  { setup TestVec }
  Series1->Clear();
  Series2->Clear();
  Series3->Clear();
  Screen->Cursor = crHourGlass;
  if (RadioGroup1->ItemIndex == 0) // { test vector }
  {
	j = 1;
	MaxIter = 10;
	if (Len < 5000) { MaxIter = 5000; }
	if (Len < 2000) { MaxIter = 10000; }
	if (Len < 500) { MaxIter = 60000; }
	if (Len < 100) { MaxIter = 300000; }

	while (j < Len) {

	  TestVec.Size(j);
	  TestVec.SetZero();
	  StartTimer();
//      { arrays property}
	  for (i = 1; i < MaxIter; i++) CumulativeSum(TestVec);
	  TimeElapsed = StopTimer()*1000;;
	  Series1->AddXY(j,TimeElapsed,"",clTeeColor);
	  Series1->Title = "Values property";
	  TestVec.SetZero();
	  StartTimer();
//      { indexed pointer }
	  for (i = 1; i < MaxIter; i++) CumulativeSumFast(TestVec);
	  TimeElapsed = StopTimer()*1000;;
	  Series2->AddXY(j,TimeElapsed,"",clTeeColor);
	  Series2->Title = "Indexed pointer";
	  if (j < 10) { j+=  2; }
	  else if (j < 100) { j+= 10; }
	  else if (j < 500) { j+= 50; }
	  else if (j < 1000) { j+= 100; }
	  else if (j < 3000) { j+= 250; }
	  else { j+=  500; }
	}
  } else {

	j = 5;
	MaxIter = 100;
	if (Len < 70) { MaxIter = 1500; }
	if (Len < 30) { MaxIter = 5500; }
	while (j < Len) {

		  TestMtx.Size(j,j);
		  TestMtx.SetVal(0); //SetZero
	//      { indexed pointer}
		  StartTimer();
		  for (i = 1; i < MaxIter; i++) CumulativeSumFast1(TestMtx);
		  TimeElapsed = StopTimer()*1000;;
		  Series2->AddXY(j,TimeElapsed,"",clTeeColor);
		  Series2->Title = "Values1D property";
	//      { 2D array property }
		  StartTimer();
		  for (i = 1; i < MaxIter; i++) CumulativeSum(TestMtx);
		  TimeElapsed = StopTimer()*1000;;
		  Series1->AddXY(j,TimeElapsed,"",clTeeColor);
		  Series1->Title = "Values 2D property";
	///      { 1D property}
		  StartTimer();
		  for (i = 1; i < MaxIter; i++) CumulativeSumFast(TestMtx);
		  TimeElapsed = StopTimer()*1000;;
		  Series3->AddXY(j,TimeElapsed,"",clTeeColor);
		  Series3->Title = "Indexed pointer";
		  if (j < 10) {j+= 2;}
		  else if (j < 50) {j+= 5;}
		  else if (j < 100) {j+= 10;}
		  else if (j < 200) {j+= 15;}
		  else {j+= 20;}
	}
  }
  Chart1->LeftAxis->Title->Caption = "time [ms] ("+IntToStr(MaxIter)+" iterations)";
  Screen->Cursor = crDefault;
}
//---------------------------------------------------------------------------

void __fastcall TDefArray::RadioGroup1Click(TObject *Sender)
{
  switch (((TRadioGroup*) Sender)->ItemIndex) {
  case 0:   Label2->Caption = "Vector length";
            TrackBar1->Max = 10000;
            TrackBar1->Frequency = 500;
            break;
  case 1:   Label2->Caption = "Sqared matrix rows";
            TrackBar1->Max = 300;
            TrackBar1->Frequency = 10;
            TrackBar1->Position = 150;
            break;
  }

  Series3->Active = (((TRadioGroup*) Sender)->ItemIndex != 0);
  TrackBar1Change(TrackBar1);
}
//---------------------------------------------------------------------------

